File: solution/f#/Nir-Kata/Primitive.Obsession/NIR.fs
namespace Nir_Kata.Primitive.Obsession

open System

module NIR =
    let private validNIRLength = 15
    let private isANumber str = str |> Seq.forall Char.IsDigit

    let private readNumber str : Option<int64> =
        if isANumber str then
            Some(str |> int64)
        else
            None

    let private validateLength (input: string, continueWith: string -> bool) =
        if input.Length = validNIRLength then
            continueWith input
        else
            false

    let private validateSex sex = sex = '1' || sex = '2'
    let private validateYear year = isANumber year

    let private validateNumber (str: string, rule: int64 -> bool) =
        match readNumber str with
        | Some n -> rule n
        | _ -> false

    let private validateMonth month =
        validateNumber (month, (fun x -> x > 0 && x <= 12))

    let private validateDepartment department =
        validateNumber (department, (fun m -> (m > 0 && m <= 95) || m = 99))

    let private validateCity city = isANumber city
    let private validateSerialNumber serialNumber = isANumber serialNumber
    let private calculateKey number : int64 = 97L - (number % 97L)

    let private validateKey number key =
        match (readNumber number, readNumber key) with
        | Some n, Some k -> calculateKey n = k
        | _ -> false

    let private validateCharacters (input: string) =
        validateSex input.[0]
        && validateYear input.[1..2]
        && validateMonth input.[3..4]
        && validateDepartment input.[5..6]
        && validateCity input.[7..9]
        && validateSerialNumber input.[10..12]
        && validateKey input.[0..12] input.[13..14]

    let validate input =
        validateLength (input, validateCharacters)

File: solution/c#/Nir-Kata/Primitive.Obsession/NIR.cs
namespace Nir_Kata.Primitive.Obsession
{
    public static class NIR
    {
        private const int ValidLength = 15;

        public static bool Validate(string input) => ValidateLength(input)
                                                     && ValidateSex(input[0])
                                                     && ValidateYear(input[1..3])
                                                     && ValidateMonth(input[3..5])
                                                     && ValidateDepartment(input[5..7])
                                                     && ValidateCity(input[7..10])
                                                     && ValidateSerialNumber(input[10..13])
                                                     && ValidateKey(input[..13], input[13..15]);

        private static bool ValidateLength(string input) => input.Length == ValidLength;

        private static bool ValidateSex(char sex) => sex is '1' or '2';
        private static bool ValidateYear(string year) => year.IsANumber();

        private static bool ValidateMonth(string month) =>
            ValidateNumber(month, m => m is > 0 and <= 12);

        private static bool ValidateDepartment(string department) =>
            ValidateNumber(department, d => d is > 0 and <= 95 or 99);

        private static bool ValidateCity(string city) => city.IsANumber();

        private static bool ValidateSerialNumber(string serialNumber) => serialNumber.IsANumber();

        private static bool ValidateNumber(string potentialNumber, Predicate<int> predicate) =>
            potentialNumber
                .ToInt()
                .Match(number => predicate(number), false);

        private static bool ValidateKey(string number, string key) =>
            number
                .ToLong()
                .Match(n => IsValidKey(n, key), false);

        private static bool IsValidKey(long number, string key) =>
            key
                .ToInt()
                .Match(k => CalculateKey(number) == k, false);

        private static long CalculateKey(long number) => 97 - (number % 97);
    }
}